Fixed a lot of parsing errors

terhechte opened the pull request:
Hey! I recently ran `email-parser` on a batch of ~650.000 emails from 2004 - 2021. Initially roughly half of those emails were marked as invalid. I slowly went through the issues and fixed one after the other in order to be able to parse as many emails a possible. All the changes can be found in this PR.
For me, it was important that I'm able to parse the majority of my mails. With these changes, all mails except for ~100 are parsed properly (and those are broken beyond repair, I had a brief look). However, I'm not sure if all of my changes are sensible additions. I added one commit for each fix so that you can easily figure out which ones are interesting. I also added tests (in the last commit) that involve most of the fixed issues.

Finally, I've added two feature flags in order to improve parsing:

- `allow-duplicate-headers`: Gmail seems to add multiple `To` fields if forwarding from one Gmail account to another. Similarly, I had many emails with multiple `Message-ID` fields and so on. I've encapsulated this into a flag. If active, multiple headers are allowed, and in case the value is a list of items (e.g. `Reply-To`) they're all merged together. If inactive, the previous behavior of not allowing multiple headers stays as before.
- `decode-mime-body`: I do need mime support for the parsing of `Subject` fields, but I'm not interested in bodies. Given that parsing of bodies might be expensive, I added a flag to specifically disable the parsing of bodies if mime is active.

I also added one dependency. This could also be made into a feature flag I guess. This library converts timezone abbreviations (such as `GMT`) to proper timezone information. I build that yesterday for the specific purpose of parsing timezone information in `email-parser`.

Cheers & Thanks for this nice library!

terhechte committed:
Added support for timezone abbreviations

terhechte committed:
Add support for the 9:23:47 time format

terhechte committed:
Fix compile issue when enabling MIME

terhechte committed:
Support for latin1, shiftjs, etc codepage chars in subjects

terhechte committed:
Support emails with no sender but multiple from addresses

terhechte committed:
Support '.' character in from names (e.g. "From: hey.io <info@heo.io>"

terhechte committed:
Support more address chars in From. Add '@'

terhechte committed:
Support for years in dates with only two digits. E.g. 11 for 2011

terhechte committed:
Support display names in quotes with any characters inside

terhechte committed:
Add support for '+00:00' timezone format

terhechte committed:
Some emails add a comment at the end of the date to indicate a timezone

terhechte committed:
Allow multiple 'to' values in headers as this is something gmail does during forwarding

terhechte committed:
Add feature flag to allow multiple headers for all fields. Vecs are joined, others ignored.

terhechte committed:
Some addresses have additional whitespace after the closing angle

terhechte committed:
Fix for an issue where unknown fields had weird unicode letters in their values

terhechte committed:
Support broken headers by moving them into unsupported (making parsing more resilient)

terhechte committed:
Add feature to disable mime body decoding

terhechte committed:
Add option to get the number for a month

terhechte committed:
Added tests for many of the issues that were fixed in the past commits

terhechte committed:
Update dependency

Mubelotix commented on email-parser/src/parsing/common.rs:
```diff
@@ -75,6 +88,16 @@ pub fn word(input: &[u8]) -> Res<Cow<str>> {
     )
 }
 
+pub fn in_quotes(input: &[u8]) -> Result<(&[u8], Vec<Cow<str>>), Error> {
```
```suggestion
pub fn in_quotes(input: &[u8]) -> Res<Vec<Cow<str>>> {
```

Mubelotix commented on email-parser/src/parsing/common.rs:
```diff
@@ -121,12 +145,30 @@ pub fn unstructured(input: &[u8]) -> Result<(&[u8], Cow<str>), Error> {
     Ok((input, output))
 }
 
+pub fn unstructured_until_linebreak(input: &[u8]) -> Result<(&[u8], Cow<str>), Error> {
```
```suggestion
pub fn unstructured_until_linebreak(input: &[u8]) -> Res<Cow<str>> {
```

Mubelotix commented on email-parser/Cargo.toml:
```diff
@@ -12,6 +12,7 @@ keywords = ["email", "mail", "mime", "parser"]
 
 [dependencies]
 textcode = {version="0.2", optional=true}
+timezone-abbreviations = "0.1.0"
```
Correct me if I am wrong but custom timezone is defined by RFC 822 and this crate is focusing on RFC 5322. There is a feature named `compatibility-fixes` to allow older syntaxes. Please put everything timezone-related under this feature gate (including the new dependency).

Mubelotix commented on email-parser/Cargo.toml:
```diff
@@ -35,6 +36,8 @@ compatibility-fixes = []
 content-disposition = ["mime"]
 unrecognized-headers = ["mime"]
 mime = ["textcode"]
+allow-duplicate-headers = []
```
I don't think we need a feature for that. I would like to leave the `Email` struct untouched, and rather create a new `PermissiveEmail` struct storing headers as Vec of their values so that it allows duplicate and even missing headers. But `Email` should designate a compliant email.

Mubelotix reviewed (changes requested):
Hey! Thank you very much for this PR! Your library crate for timezones is nice!
I'm quite surprised by the share of invalid emails in your dataset. Is this a public dataset I could download somewhere?
I made some review comments mostly about cargo features. The thing is the performance of parsing valid emails should not be affected by the workarounds required by non-compliant emails unless the `compatibility-fixes` is explicitly enabled.

terhechte commented:
> Hey! Thank you very much for this PR! Your library crate for timezones is nice!
> I'm quite surprised by the share of invalid emails in your dataset. Is this a public dataset I could download somewhere?
> I made some review comments mostly about cargo features. The thing is the performance of parsing valid emails should not be affected by the workarounds required by non-compliant emails unless the `compatibility-fixes` is explicitly enabled.

Ah no, these are my personal emails. I did a gmail download with the intention of generating statistics on them. I can probably dig out a couple of the worst offenders and remove some personal information from them (e.g. the body) to share.

Somehow I missed the compatibility-fixes feature flag. I suppose it would be nice to have a brief explanation of all the existing feature flags. I also wasn't sure what `keywords` and `trace` do.

Regarding the `PermissiveEmail` I see the appeal of that. I'll see how much time I can invest to implement this. Right now the parser works good enough for the problem I'm trying to solve so I'll probably focus on that first and afterwards look into this PR again :)

Mubelotix commented:
> Somehow I missed the compatibility-fixes feature flag. I suppose it would be nice to have a brief explanation of all the existing feature flags. I also wasn't sure what keywords and trace do.

That would indeed be nice. `keywords` enables parsing for the keywords header and `trace` was supposed to parse all the trace-related headers. Unfortunately, the RFC is insanely vague and wrong about these headers. So I have no idea how I am supposed to implement them.

> Regarding the PermissiveEmail I see the appeal of that. I'll see how much time I can invest to implement this. Right now the parser works good enough for the problem I'm trying to solve so I'll probably focus on that first and afterwards look into this PR again :)

I will do it if you can't :)

terhechte committed:
Support Apple Mail messages which use LF instead of CRLF

I've moved this under compatibility-fixes, but maybe that could
also move under its own feature flag for emlx
